### ISSUE TITLE
Dilated Pooling

### ISSUE THREAD

**Issue by f0k** [Theano/Theano#6002]
A CNN with pooling layers trained on small patches will produce an output of reduced resolution when applied to larger patches (reduced by the product of all pooling factors involved). With dilated convolution and dilated pooling it can be efficiently applied to large patches without reducing output resolution.
(Specifically, one would set all pooling layers to a stride of 1, and set the dilation of all convolutional and pooling layers to the product of all pooling factors encountered before. See e.g. http://arxiv.org/abs/1412.7062, Section 3.1.)

This would need an extension of the pooling interface, and an implementation -- as a first step, it can be implemented as a graph optimization padding the input to be divisible by the dilation factors, reshaping to fold the dilation offsets into the batch dimension, applying standard pooling and reshaping back. A second step would be explicit support. Depending on how the pooling switches for the backward pass are stored, it's possible that only the forward pass needs to change.

Torch has an implementation here: https://github.com/torch/cunn/blob/master/lib/THCUNN/SpatialDilatedMaxPooling.cu
Caffe has an implementation in its opencl branch: https://github.com/BVLC/caffe/blob/opencl/src/caffe/layers/pooling_layer.cu
TF also implements it: https://www.tensorflow.org/api_docs/python/tf/nn/pool

**Comment by lamblin** [comment]
I think I'd need some time to wrap my head around that, I'm not sure what dilated pooling does exactly.

Currently, the pooling switches are implicitly stored as pooling inputs and pooling outputs, and the places where input == output are the locations where to backprop.

**Comment by f0k** [comment]
> I think I'd need some time to wrap my head around that, I'm not sure what dilated pooling does exactly.

It's basically the same connection pattern as with dilated convolution. I.e., dilated mean-pooling is the same as dilated convolution with all filter elements set to a suitable constant (i.e., 1 / (pool_height * pool_width)). Its main use is for applying an existing network with multiple pooling layers (e.g., AlexNet, VGG, ResNet, ...) to get a prediction for every pixel of an image.

> Currently, the pooling switches are implicitly stored as pooling inputs and pooling outputs, and the places where input == output are the locations where to backprop.

I see. The `input == output` comparison has to consider the pooling window locations then. This means the backward pass has to take dilation into account as well.

**Comment by lamblin** [comment]
OK, I think I've got it.
In that case, I think we only need dilation > 1 with a stride == 1. Or should we worry about combining the two as well?

**Comment by f0k** [comment]
> In that case, I think we only need dilation > 1 with a stride == 1.

Yes, that's right.

> Or should we worry about combining the two as well?

That's probably not needed. The first use case that came to my mind for this would be applying VGG to get a prediction for every third pixel of an image, but that would be done by changing the stride of the first pooling layer, which is never dilated.
In any case, I'd suggest to add `dilation` to the existing pooling interface rather than creating a new interface for dilated convolution that does not support strides. It's okay if not all cases are implemented. But we could borrow the caffe kernel (the one also used by torch) for a fall-back GPU implementation that supports all combinations.

**Comment by rcasero** [comment]
I have requested this feature in TensorFlow

https://github.com/tensorflow/tensorflow/issues/12681

and Keras, as I'd like to test whether it's better (it's used by DeepCell for cell segmentation)

https://groups.google.com/forum/#!topic/keras-users/ZVtI8Ef6508

(@fchollet commented that they'd be happy to modify the Keras API to add dilated pooling once it's implemented in low-level).

However, @f0k , it's not dilation what deals with resolution loss, is it? That's a matter for stride. So if you have non-dilated pooling with a stride of 1, then you won't lose resolution anyway, right?

**Comment by f0k** [comment]
> So if you have non-dilated pooling with a stride of 1, then you won't lose resolution anyway, right?

Yes, but then you'd have a different model. The use case here was to take a model that was trained with strided pooling, and then apply it with stride 1 to get a full-resolution output. For this all convolutions and poolings *after* the (previously strided, now non-strided) pooling layer have to be dilated, so they basically see stride-pooled features again.
